Add OPA blog post (#3953)
pulumi/docs: content/blog/opa-support-for-crossguard/index.md

To test out our rule, we can create a new Pulumi Kubernetes project and deploy it with the policy enforced.

First we create a new project `kubeapp` in a sibling folder:

```
$ cd .. && mkdir kubeapp && cd kubeapp
$ pulumi new kubernetes-typescript
```

This will create a project that deploys a single Kubenetes `Deployment` (but without our required labels!).

```ts
import * as k8s from "@pulumi/kubernetes";
import * as kx from "@pulumi/kubernetesx";

const appLabels = { app: "nginx" };
const deployment = new k8s.apps.v1.Deployment("nginx", {
    spec: {
        selector: { matchLabels: appLabels },
        replicas: 1,
        template: {
            metadata: { labels: appLabels },
            spec: { containers: [{ name: "nginx", image: "nginx" }] }
        }
    }
});
export const name = deployment.metadata.name;
```

> We can deploy this to any Kubernetes cluster (just configure your `kubectl` context locally to point at the desired cluster) with our policy applied with:

Reviewer: ```suggestion
We can deploy this to any Kubernetes cluster (configure your `kubectl` context locally to point at the desired cluster) with our policy applied with:
```

```
$ pulumi up --policy-pack ../policy-kubernetes
Previewing update (dev):
     Type                           Name                   Plan       Info
 +   pulumi:pulumi:Stack            simple-kubernetes-dev  create     1 error
 +   └─ kubernetes:apps:Deployment  nginx                  create

Diagnostics:
  pulumi:pulumi:Stack (simple-kubernetes-dev):
    error: preview failed

Policy Violations:
    [mandatory]  kubernetes v0.0.1  deny (nginx: kubernetes:apps/v1:Deployment)
    nginx-me0llhgr must include Kubernetes recommended labels: https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/#labels
```

> Note that the preview fails because the policy was violated, preventing me from attempting to deploy this resource to my cluster!

Reviewer: ```suggestion
Note that the preview fails because the policy was violated, preventing me from deploying this resource to my cluster!
```

> The policy defined above has a "mandatory" policy, defined in OPA with `deny`. We can also define "advisory" policies using `warn` rules. For example, if we cahnged our policy from `deny` to `warn` and rerun - we will see the preview succeed:

Reviewer: ```suggestion
The policy defined above has a "mandatory" policy, defined in OPA with `deny`. We can also define "advisory" policies using `warn` rules. For example, if we changed our policy from `deny` to `warn` and rerun - we will see the preview succeed:
```

```
$ pulumi up
Previewing update (dev2):
Installing policy pack kubernetes 2...
Finished installing policy pack

     Type                           Name                    Plan
 +   pulumi:pulumi:Stack            simple-kubernetes-dev2  create
 +   └─ kubernetes:apps:Deployment  nginx                   create

Policy Violations:
    [advisory]  kubernetes v0.0.1  warn (nginx: kubernetes:apps/v1:Deployment)
    nginx-q4ilsuqq must include Kubernetes recommended labels: https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/#labels

Do you want to perform this update? yes
Updating (dev2):
     Type                           Name                    Status
 +   pulumi:pulumi:Stack            simple-kubernetes-dev2  created
 +   └─ kubernetes:apps:Deployment  nginx                   created

Policy Violations:
    [advisory]  kubernetes v0.0.1  warn (nginx: kubernetes:apps/v1:Deployment)
    nginx-6ooayw4j must include Kubernetes recommended labels: https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/#labels

Outputs:
    name: "nginx-6ooayw4j"

Permalink: https://app.pulumi.com/lukehoban/simple-kubernetes/dev2/updates/1
```

> We can now navigate to the URL for our update and see that the policy was enforced on this update. Note the badge noting that this deployment was protected by policy enforcement, and the list of policy violations.

Reviewer: ```suggestion
We can now navigate to the URL for our update and see that the policy was enforced. Note the badge showing that this deployment was protected by policy enforcement and the list of policy violations.
```

![Pulumi Console Policy Enforcement](./console-policy-enforcement.png)

## Using OPA Rego with Pulumi

> OPA-based rules are easy to write for Pulumi. The `input` object represents each individual resource that will be deployed by Pulumi, and the rules are applied independently to each Pulumi resource when it is deployed. The schema for each resource is the same as the properties availble when defining the resource as defined in the [Pulumi API Reference](https://www.pulumi.com/docs/reference/pkg/). Critically, although we higlighted Kubernetes above, Pulumi's resource model spans any cloud or Kuberntes, with providers for more than 50 Cloud and SaaS platforms available. And OPA-based rules can target any or all of these in a single deployment.

Reviewer: ```suggestion
OPA-based rules are easy to write for Pulumi. The `input` object represents each resource deployed by Pulumi, and the rules are applied independently to each Pulumi resource when it is deployed. Each resource's schema is the same as the properties available when defining the resource, as specified in the [Pulumi API Reference]({{< relref "/docs/reference/pkg" >}}). Critically, although we highlighted Kubernetes above, Pulumi's resource model spans any cloud or Kubernetes, with providers for more than 50 Cloud and SaaS platforms available. And OPA-based rules can target any or all of these in a single deployment.
```

For example - the policy file above could be expanded to include:

```rego
# Enforce that AWS EC2 Instance public IP is not enabled
warn[msg] {
  input.associatePublicIpAddress
  msg := "Associating public IP addresses is discouraged"
}

# Enforce that Kubernetes Pod images must be published by `hooli.com`
deny[msg] {
  input.kind == "Pod"
  image := input.spec.containers[_].image
  not startswith(image, "hooli.com/")
  msg := sprintf("image '%v' comes from untrusted registry", [image])
}
```

With support in Visual Studio Code, authoring both your infrastructure and policy in a rich IDE using Pulumi is also possible.

![VSCode for OPA and Pulumi](./vscode-rego.png)

## Conclusion

> OPA support in Pulumi Crossguard enables authoring cloud infrastructure policy using a familiar and popular policy authoring language while leveraging Pulumi's Crossguard platform for policy enforcement from development to production. Along with existing support for TypeScript, JavaScript and Pythong - Pulumi CrossGuard offers a great breadth of authoring languages to meet the needs of any infrastructure team.

Reviewer: ```suggestion
OPA support in Pulumi Crossguard enables authoring cloud infrastructure policy using a familiar and popular policy authoring language while leveraging Pulumi's Crossguard platform for policy enforcement from development to production. Along with existing support for TypeScript, JavaScript, and Python, Pulumi CrossGuard offers a great breadth of authoring languages to meet any infrastructure team's needs.
```

> Check out the open source [pulumi-policy-opa](https://github.com/pulumi/pulumi-policy-opa) project to learn more about the OPA integration with CrossGuard. And [get started](https://www.pulumi.com/docs/get-started/) with Pulumi for your modern cloud infrastructure and policy today.

Reviewer: ```suggestion
Check out the open source [pulumi-policy-opa](https://github.com/pulumi/pulumi-policy-opa) project to learn more about the OPA integration with CrossGuard. And [get started]({{< relref "/docs/get-started" >}}) with Pulumi for your modern cloud infrastructure and policy today.
```
